core/src/main/java/io/grpc/internal/ServerCallImpl.java

```diff
@@ -327,16 +327,20 @@ private void messagesAvailableInternal(final MessageProducer producer) {
         return;
       }
 
-      InputStream message;
       try {
+        InputStream message;
         while ((message = producer.next()) != null) {
-          try {
-            listener.onMessage(call.method.parseRequest(message));
-          } catch (Throwable t) {
+          ReqT parsedMessage;
+          try (InputStream ignored = message) {
+            parsedMessage = call.method.parseRequest(message);
+          } catch (StatusRuntimeException e) {
             GrpcUtil.closeQuietly(message);
-            throw t;
+            GrpcUtil.closeQuietly(producer);
+            call.cancelled = true;
+            call.close(e.getStatus(), new Metadata());
```

Comment:
This breaks the API pretty solidly.

call.close() is the public API, and it has `closeCalled` as protection to make sure the API is used correctly. The application won't know about this close() you're calling here, so the application will very possibly end up calling close() themselves and then getting an incorrect exception.

Also, the application won't be notified the RPC was cancelled; yes isCancelled returns true, but onCancel isn't called. onComplete() is also called eventually which means all the server's messages were sent, but they weren't.